libiprange/iprange.3
.TH libiprange 3
.SH NAME
.B iprange_init(), iprange_close(), iprange_geterror(), iprange_setdomain(),
.B iprange_query_start(), iprange_query_check(), iprange_query_cancel(),
.B iprange_settimeout(), iprange_setcallbackint(), iprange_setcallbackctx(),
.B iprange_setdnscallback(), iprange_dns_set_query_service(),
.B iprange_dns_set_query_start(), iprange_dns_set_query_cancel(),
.B iprange_dns_set_query_waitreply()
-- range-based Real-time Block List (RBL) query service
.SH SYNOPSIS
#include
.I <iprange.h>

RBL *
.B iprange_init
(void *(*
.B mallocf
)(void *, size_t), void (*
.B freef
)(void *, void *p), void *
.B closure
);

unsigned char *
.B iprange_geterror
(RBL *
.B iprange
);

void
.B iprange_setdomain
(RBL *
.B iprange,
unsigned char *
.B qroot
);

IPRANGE_STAT
.B iprange_query_start
(RBL *
.B iprange,
unsigned char *
.B query,
void **
.B qh
);

IPRANGE_STAT
.B iprange_settimeout
(RBL *
.B iprange,
unsigned int
.B timeout
);

IPRANGE_STAT
.B iprange_setcallbackint
(RBL *
.B iprange,
unsigned int
.B cbint
);

IPRANGE_STAT
.B iprange_setcallbackctx
(RBL *,
.B iprange,
void *
.B ctx
);

IPRANGE_STAT
.B iprange_setdnscallback
(RBL *,
.B iprange,
void (*
.B func
)(const void *));

void *
.B iprange_dns_set_query_service
(RBL *
.B iprange,
void *
.B svc
);

void
.B iprange_dns_set_query_cancel
(RBL *
.B iprange,
int (*
.B func
)(void *, void *));

void
.B iprange_dns_set_query_start
(RBL *
.B iprange,
int (*
.B func
)(void *, int, unsigned char *, unsigned char *, size_t, void **));

void
.B iprange_dns_set_query_waitreply
(RBL *
.B iprange,
int (*
.B func
)(void *, void *, struct timeval *, size_t *, int *, int *));

void
.B iprange_close
(RBL *
.B iprange
);

.SH DESCRIPTION
These functions are an interface to a facility to conduct Realtime Blacklist
(RBL) queries that use the draft specification for IP range publications,
and return their results.  Basic RBLs are described in RFC5782; range-based
RBLs are defined in draft-levine-iprangepub.

An application first initializes the package by calling
.B iprange_init().
The optional
.I mallocf
parameter is a caller-provided memory allocation function taking a pointer
to a caller-provided opaque data structure (the
.I closure
parameter) and a number of bytes to allocate.  If
.I mallocf
is not provided, the default system memory allocator function
.B malloc(3)
is used.  The optional
.I freef
parameter specifies a matching caller-provided memory deallocator function,
taking a pointer to a caller-provided opaque data structure (the
.I closure
parameter) and a pointer to the memory to be released.  If
.I freef
is not provided, the default system memory release function
.B free(3)
is used.  A handle for future use of the initialized library instance
is returned, or NULL on error and
.I errno
will be set to indicate the cause of the failure.

The
.B iprange_geterror()
function can be used to poll the library for an error string that provides
further description for the most recent failed operation.

A call to
.B iprange_setdomain()
sets the RBL's base domain.  This specifies the "root" of all future queries
for the provided
.B iprange
handle to
.B qroot.
Future queries passed to this handle will be prepended to the string
provided here.

A query is initialized with a call to
.B iprange_query_start()
which takes an IP range library handle as
.B iprange,
the string to be queried as
.B query,
and a pointer to a void pointer that will be updated to contain a reference
to the initiated query as
.B qh.

A query previously started by
.I iprange_query_start()
can be canceled by calling
.B iprange_query_cancel(),
passing it an IP range library handle as
.B iprange
and a query handle as 
.B qh
as previously returned when the query was started.

A query previously started by
.I iprange_query_start()
can be polled for its status or a result using
.B iprange_query_check(),
passing it an IP range library handle as
.B iprange,
a query handle as 
.B qh,
a pointer to a "struct timeval" ast
.B timeout
(which can be NULL if an infinite wait is acceptable), and a pointer to
a 32-bit integer as
.B res
into which any result will be returned (or it can be NULL if the caller
doesn't actually want the result).  Results are in network byte order.
If this function returns IPRANGE_STAT_FOUND or IPRANGE_STAT_NOTFOUND, the query
is automatically canceled and need not be passed to
.I iprange_query_cancel().

The
.B iprange_settimeout()
function can be used to change the query timeout.  The default is ten seconds.
This applies to each DNS activity individually, and a single query through
this interface may be made up of more than one such query.

If it is useful to have the library periodically call a user-provided function
as an indication that queries are still in progress, such a function can be
registered with the
.B iprange_setdnscallback()
function.  The function provided should take a void context pointer as
its sole argument.
.B iprange_setcallbackctx()
is used to tell the library what context pointer should be used, and
.B iprange_setcallbackint()
is used to tell the library what frequency, in seconds, should be used to
call that callback function.  Each of these takes the corresponding RBL
library handle as its first argument, and the obvious parameter for its second.

By default, the library will use the stock system resolver to conduct DNS
queries.  If alternates should be used, these can be specifed using the
following functions:

.B iprange_dns_set_query_service()
sets a context pointer to the query service to be used, if any.  This will
be passed as-is to the other DNS functions.  It returns its previous value.

.B iprange_dns_set_query_cancel()
sets a pointer to the function that should be called to cancel an open query,
and should take the following parameters: a void pointer to the DNS query
service to be used, and a void pointer to a query handle as previously returned
by a call to
.I iprange_dns_set_query_start().
The function should return one of the DNS result codes described below.

.B iprange_dns_set_query_start()
sets a pointer to the function that should be called to initiate a new query,
and should take the following parameters: a void pointer to the DNS query
service to be used, a DNS record type (e.g. T_TXT), a pointer to a string
containing the query to be started, a pointer to a buffer into which the
reply should be written, the number of bytes available in that buffer,
and a pointer to a void pointer that will be updated to contain a unique
handle for that query once started.  The function should return one of the
DNS result codes described below.

.B iprange_dns_set_query_waitreply()
sets a pointer to the function that should be called to wait for a reply
to an open query, and should take the following parameters: a void pointer
to the DNS query service to be used, a void pointer referencing the query of
interest as previously returned by a call to
.I iprange_dns_set_query_start(),
a pointer to a "struct timeval" structure indicating how long the function
should wait for a reply (or NULL if infinite waiting is acceptable),
a pointer to a "size_t" that will be updated to contain the size of the
received reply, a pointer to an integer that will contain an error code
if the query fails (can be NULL if that information is not interesting to
the caller), and a pointer to an integer that will contain a DNSSEC
status indication (can be NULL if that information is not interesting to
the caller).  The function should return one of the DNS result codes
described below.

When the library handle is no longer needed, it should be passed to
.B iprange_close().

.SH RETURN VALUES
The following return codes, of type IPRANGE_STAT, can be returned:
.TP
.I IPRANGE_STAT_OK
successful completion
.TP
.I IPRANGE_STAT_INVALID
operation failed because an invalid parameter was provided
.TP
.I IPRANGE_STAT_DNSERROR
operation could not be completed because of errors requesting or receiving
a DNS reply; note that this does not include a successful reply that contains

RELEASE_NOTES
			OPENDKIM RELEASE NOTES
	$Id: RELEASE_NOTES,v 1.258 2010/10/25 17:17:46 cm-msk Exp $


This listing shows the versions of the OpenDKIM package, the date of
release, and a summary of the changes in that release.

2.4.0		2011/06/06
	Feature request #SF3060140: Extension of odkim.sign() to include
		maximum signed byte count.
	Feature request #SF3171865: Add _FFR_LUA_GLOBALS, enabling a
		mechanism to maintain state between Lua scripts.
	Feature request #SF3241982: Optionally an Authentication-Results
		field for all signatures present, not just the first good
		one.
	Feature request #SF3256630: Add _FFR_RATE_LIMIT, enabling hooks for
		post-verification rate controls.
	Feature request #SF3290461: Add optional support for querying a
		memcache daemon via libmemcached.
	Feature request #SF3299041: Be tougher on use of the "l=" tag by
		removing "BodyLengths", renaming "BodyLengthDBFile" to
		"BodyLengthDB", and allowing pattern matches.  This
		effectively activates _FFR_BODYLENGTH_DB as well.
	LIBOPENDKIM: Feature request #SF2969516: Teach dkim_chunk() about
		DKIM_LIBFLAGS_FIXCRLF.
	LIBOPENDKIM: Feature request #SF3123731: Extend FixCRLF to cover
		header fields.
	LIBOPENDKIM: Feature request #SF3151534: Extract DKIM reputation
		support to its own library.
	LIBOPENDKIM: Feature request #SF3279589: Add _FFR_XTAGS, adding
		support for generic extension tags in signatures.
	LIBOPENDKIM: Feature request #SF3300738: Add dkim_sig_getqueries()
		and dkim_policy_getqueries() to assist API consumers that
		have their own DNS mechanisms not directly adaptable to ours.
		Requested by Heikki Gruner.
	STATS: Feature request #SF3066106: Add "opendkim-anonstats" tool.
	CONTRIB: Paths in init/redhat/opendkim are now auto-generated.
	Activate _FFR_MAXVERIFY.

2.3.2		2011/04/14
	Fix bug #SF3067338: Create a pool of DB handles for SQL operations
		so that queries don't all share one serially.
	Fix bug #SF3156559: Don't compile Lua scripts anew for each execution.
	Fix bug #SF3210385: Fix issue with configuration reload failures
		having some side effects on the remaining active configuration
		handle.  Reported by Andreas Schulze.
	Fix bug #SF3235131: Don't report ADSP errors as signature errors
		in Authentication-Results header fields.
	Fix bug #SF3258459: Don't send progress messages during EOH.  Problem
		reported by Gary Mills.
	Add ADSP processing error information to A-R header fields.
	LIBOPENDKIM: Fix bug #SF3276469: When generating "z=" tags, apply
		omit and sign header field lists.  Problem noted by Todd Lyons.
	LIBOPENDKIM: Use a time_t to store TTL and clock drift options
		rather than unsigned long.  Patch from Piotr Sikora.
	LIBAR: Include class and type in debugging messages.  Suggested
		by Gary Mills.
	LIBAR: Start the dispatcher on the first query, not in ar_init().
		Allows ar_init() to be called before a fork() safely.
	CONTRIB: Adjust installed path in RedHat start script.  Patch
		from Naresh V.
	CONTRIB: Require the current libopendkim in the opendkim definition
		in the RPM spec file.  Patch from Naresh V.

2.3.1		2011/03/28
	Feature request #SF3196060: Log retrieved reputation values.
		Patch from Andreas Schulze.
	Feature request #SF3199824: Don't log "no signature data" unless
		LogWhy is enabled or the action for unsigned mail is something
		other than "accept".
	Fix bug #SF3196069: Log more detail when dkimf_config_setlib() fails.
		Problem noted by Andreas Schulze.
	Fix bug #SF3197890: Update ARF report format for DKIM verification
		failures to use what's in draft-ietf-dkim-marf-reporting.
	Fix bug #SF3204725: Add "ReplaceHeaders" data set to constrain what
		header fields are affected by "ReplaceRules".  Problem
		noted by Chris Callegari.
	Fix bug #SF3211114: Extract "BaseDirectory" from configuration
		file properly.
	Fix bug #SF3234178: Correct handling of a KeyTable entry for which the
		key data is empty.  Based on a patch from Yoshiaki Yanagihara.
	Patch #SF3205699: Make all header field name and domain name data

libiprange/iprange.c
/*
**  Copyright (c) 2010, 2011, The OpenDKIM Project.  All rights reserved.
*/

/* TODO:
** - needs a worker pool that handles all of the various subqueries
**   for a given main query, so the caller only waits on a final result
** - needs support for IPv6
*/

#ifndef lint
static char iprange_c_id[] = "$Id$";
#endif /* !lint */

#include "build-config.h"

/* system includes */
#include <sys/param.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <arpa/nameser.h>
#include <netdb.h>
#include <resolv.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <errno.h>

/* libiprange includes */
#include "iprange.h"

/* local definitions needed for DNS queries */
#define MAXPACKET		8192
#if defined(__RES) && (__RES >= 19940415)
# define RES_UNC_T		char *
#else /* __RES && __RES >= 19940415 */
# define RES_UNC_T		unsigned char *
#endif /* __RES && __RES >= 19940415 */
#ifndef T_RRSIG
# define T_RRSIG		46
#endif /* ! T_RRSIG */

/* struct iprange_query -- an open IPRANGE IPRANGE query */
struct iprange_query
{
	void *			rq_qh;
	size_t			rq_anslen;
	u_char			rq_buf[HFIXEDSZ + MAXPACKET];
};

/* struct iprange_handle -- an IPRANGE IPRANGE library context */
struct iprange_handle
{
	u_int			iprange_timeout;
	u_int			iprange_cbint;
	void *			iprange_cbctx;
	void *			iprange_closure;
	void *			(*iprange_malloc) (void *closure,
				                   size_t nbytes);
	void			(*iprange_free) (void *closure, void *p);
	void			(*iprange_dns_callback) (const void *context);
	void *			iprange_dns_service;
	int			(*iprange_dns_start) (void *srv, int type,
				                      unsigned char *query,
				                      unsigned char *buf,
				                      size_t buflen,
				                      void **qh);
	int			(*iprange_dns_cancel) (void *srv, void *qh);
	int			(*iprange_dns_waitreply) (void *srv,
				                          void *qh,
				                          struct timeval *to,
				                          size_t *bytes,
				                          int *error,
				                          int *dnssec);
	u_char			iprange_qroot[IPRANGE_MAXHOSTNAMELEN + 1];
	u_char			iprange_error[IPRANGE_MAXERRORSTRING + 1];
};

/*
**  Standard UNIX resolver stub functions
*/

struct iprange_res_qh
{
	int		rq_error;
	size_t		rq_buflen;
};

/*
**  IPRANGE_RES_CANCEL -- cancel a pending resolver query
**
**  Parameters:
**  	srv -- query service handle (ignored)
**  	qh -- query handle (ignored)
**
**  Return value:
**  	0 on success, !0 on error
**
**  Notes:
**  	The standard UNIX resolver is synchronous, so in theory this can
**  	never get called.  We have not yet got any use cases for one thread
**  	canceling another thread's pending queries, so for now just return 0.
*/

static int
iprange_res_cancel(void *srv, void *qh)
{
	if (qh != NULL)
		free(qh);

	return 0;
}

/*
**  IPRANGE_RES_QUERY -- initiate a DNS query
**
**  Parameters:
**  	srv -- service handle (ignored)
**  	type -- RR type to query
**  	query -- the question to ask
**  	buf -- where to write the answer
**  	buflen -- bytes at "buf"
** 	qh -- query handle, used with iprange_res_waitreply
**
**  Return value:
**  	An IPRANGE_DNS_* constant.
**
**  Notes:
**  	This is a stub for the stock UNIX resolver (res_) functions, which
**  	are synchronous so no handle needs to be created, so "qh" is set to
**  	"buf".  "buf" is actually populated before this returns (unless
**  	there's an error).
*/

static int
iprange_res_query(void *srv, int type, unsigned char *query,
                  unsigned char *buf, size_t buflen, void **qh)
{
	int n;
	int ret;
	struct iprange_res_qh *rq;
	unsigned char qbuf[HFIXEDSZ + MAXPACKET];
#ifdef HAVE_RES_NINIT
	struct __res_state statp;
#endif /* HAVE_RES_NINIT */

#ifdef HAVE_RES_NINIT
	memset(&statp, '\0', sizeof statp);
	res_ninit(&statp);
#endif /* HAVE_RES_NINIT */

#ifdef HAVE_RES_NINIT
	n = res_nmkquery(&statp, QUERY, (char *) query, C_IN, type, NULL, 0,
	                 NULL, qbuf, sizeof qbuf);
#else /* HAVE_RES_NINIT */
	n = res_mkquery(QUERY, (char *) query, C_IN, type, NULL, 0, NULL, qbuf,
	                sizeof qbuf);
#endif /* HAVE_RES_NINIT */
	if (n == (size_t) -1)
	{
#ifdef HAVE_RES_NINIT
		res_nclose(&statp);
#endif /* HAVE_RES_NINIT */
		return IPRANGE_DNS_ERROR;
	}

#ifdef HAVE_RES_NINIT
	ret = res_nsend(&statp, qbuf, n, buf, buflen);
#else /* HAVE_RES_NINIT */
	ret = res_send(qbuf, n, buf, buflen);
#endif /* HAVE_RES_NINIT */
	if (ret == -1)
	{
#ifdef HAVE_RES_NINIT
		res_nclose(&statp);
#endif /* HAVE_RES_NINIT */
		return IPRANGE_DNS_ERROR;
	}

#ifdef HAVE_RES_NINIT
	res_nclose(&statp);
#endif /* HAVE_RES_NINIT */

	rq = (struct iprange_res_qh *) malloc(sizeof *rq);
	if (rq == NULL)
		return IPRANGE_DNS_ERROR;

	if (ret == -1)
	{
		rq->rq_error = errno;
		rq->rq_buflen = 0;
	}
	else
	{
		rq->rq_error = 0;
		rq->rq_buflen = (size_t) ret;
	}

	*qh = (void *) rq;

	return IPRANGE_DNS_SUCCESS;
}

/*
**  IPRANGE_RES_WAITREPLY -- wait for a reply to a pending query
**
**  Parameters:
**  	srv -- service handle
**  	qh -- query handle
**  	to -- timeout
**  	bytes -- number of bytes in the reply (returned)
**  	error -- error code (returned)
**
**  Return value:
**  	A IPRANGE_DNS_* code.
**
**  Notes:
**  	Since the stock UNIX resolver is synchronous, the reply was completed
** 	before iprange_res_query() returned, and thus this is almost a no-op.
*/

int
iprange_res_waitreply(void *srv, void *qh, struct timeval *to, size_t *bytes,
                      int *error, int *dnssec)
{
	struct iprange_res_qh *rq;

	assert(qh != NULL);

	rq = qh;

	if (bytes != NULL)
		*bytes = rq->rq_buflen;
	if (error != NULL)
		*error = rq->rq_error;

	return IPRANGE_DNS_SUCCESS;
}

/*
**  IPRANGE_INIT -- initialize an IPRANGE IPRANGE handle
**
**  Parameters:
**  	caller_mallocf -- caller-provided memory allocation function
**  	caller_freef -- caller-provided memory release function
**  	closure -- memory closure to pass to the above when used
**
**  Return value:
**  	A new IPRANGE RBL handle suitable for use with other IPRANGE RBL
**  	functions, or NULL on failure.
**  
**  Side effects:
**  	Sudden changes in local density altitude.
*/

IPRANGE *
iprange_init(void *(*caller_mallocf)(void *closure, size_t nbytes),
             void (*caller_freef)(void *closure, void *p),
             void *closure)
{
	IPRANGE *new;

	if (caller_mallocf == NULL)
		new = (IPRANGE *) malloc(sizeof(struct iprange_handle));
	else
		new = caller_mallocf(closure, sizeof(struct iprange_handle));

	if (new == NULL)
		return NULL;

	memset(new, '\0', sizeof(struct iprange_handle));

	new->iprange_timeout = IPRANGE_DEFTIMEOUT;
	new->iprange_closure = closure;
	new->iprange_malloc = caller_mallocf;
	new->iprange_free = caller_freef;
	new->iprange_dns_start = iprange_res_query;
	new->iprange_dns_waitreply = iprange_res_waitreply;
	new->iprange_dns_cancel = iprange_res_cancel;

	return new;
}

/*
**  IPRANGE_CLOSE -- shut down a IPRANGE instance
**
**  Parameters:
**  	iprange -- IPRANGE handle to shut down
**
**  Return value:
**  	None.
*/

void
iprange_close(IPRANGE *rbl)
{
	assert(rbl != NULL);

	if (iprange->iprange_free != NULL)
		iprange->iprange_free(iprange->iprange_closure, rbl);
	else
		free(rbl);
}

/*
**  IPRANGE_GETERROR -- return any stored error string from within the IPRANGE
**                  context handle
**
**  Parameters:
**  	iprange -- IPRANGE handle from which to retrieve an error string
**
**  Return value:
**  	A pointer to the stored string, or NULL if none was stored.
*/

const u_char *
iprange_geterror(IPRANGE *rbl)
{
	assert(rbl != NULL);

	return iprange->iprange_error;
}

/*
**  IPRANGE_SETDOMAIN -- declare the IPRANGE's domain (the query root)
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	qroot -- query root
**
**  Return value:
**  	None (yet).
*/

void
iprange_setdomain(IPRANGE *rbl, u_char *qroot)
{
	assert(rbl != NULL);
	assert(qroot != NULL);

	strncpy(iprange->iprange_qroot, qroot, sizeof iprange->iprange_qroot);
	iprange->iprange_qroot[sizeof iprange->iprange_qroot - 1] = '\0';
}

/*
**  IPRANGE_SETTIMEOUT -- set the DNS timeout
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	timeout -- requested timeout (seconds)
**
**  Return value:
**  	None.
*/

void
iprange_settimeout(IPRANGE *rbl, u_int timeout)
{
	assert(rbl != NULL);

	iprange->iprange_timeout = timeout;
}

/*
**  IPRANGE_SETCALLBACKINT -- set the DNS callback interval
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	cbint -- requested callback interval (seconds)
**
**  Return value:
**  	None.
*/

void
iprange_setcallbackint(IPRANGE *rbl, u_int cbint)
{
	assert(rbl != NULL);

	iprange->iprange_cbint = cbint;
}

/*
**  IPRANGE_SETCALLBACKCTX -- set the DNS callback context
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	ctx -- context to pass to the DNS callback
**
**  Return value:
**  	None.
*/

void
iprange_setcallbackctx(IPRANGE *rbl, void *ctx)
{
	assert(rbl != NULL);

	iprange->iprange_cbctx = ctx;
}

/*
**  IPRANGE_SETDNSCALLBACK -- set the DNS wait callback
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	func -- function to call; should take an opaque context pointer
**
**  Return value:
**  	None.
*/

void
iprange_setdnscallback(IPRANGE *rbl, void (*func)(const void *context))
{
	assert(rbl != NULL);

	iprange->iprange_dns_callback = func;
}

/*
**  IPRANGE_DNS_SET_QUERY_SERVICE -- stores a handle representing the DNS
**                               query service to be used, returning any
**                               previous handle
**
**  Parameters:
**  	iprange -- IPRANGE library handle
**  	h -- handle to be used
**
**  Return value:
**  	Previously stored handle, or NULL if none.
*/

void *
iprange_dns_set_query_service(IPRANGE *rbl, void *h)
{
	void *old;

	assert(rbl != NULL);

	old = iprange->iprange_dns_service;

	iprange->iprange_dns_service = h;

	return old;
}

/*
**  IPRANGE_DNS_SET_QUERY_START -- stores a pointer to a query start function
**
**  Parameters:
**  	iprange -- IPRANGE library handle
**  	func -- function to use to start queries
**
**  Return value:
**  	None.
**
**  Notes:
**  	"func" should match the following prototype:
**  		returns int (status)
**  		void *dns -- receives handle stored by
**  		             iprange_dns_set_query_service()
**  		int type -- DNS RR query type (C_IN assumed)
**  		char *query -- question to ask
**  		char *buf -- buffer into which to write reply
**  		size_t buflen -- size of buf
**  		void **qh -- returned query handle
*/

void
iprange_dns_set_query_start(IPRANGE *rbl, int (*func)(void *, int,
                                              unsigned char *,
                                              unsigned char *,
                                              size_t, void **))
{
	assert(rbl != NULL);

	iprange->iprange_dns_start = func;
}

/*
**  IPRANGE_DNS_SET_QUERY_CANCEL -- stores a pointer to a query cancel function
**
**  Parameters:
**  	iprange -- IPRANGE library handle
**  	func -- function to use to cancel running queries
**
**  Return value:
**  	None.
**
**  Notes:
**  	"func" should match the following prototype:
**  		returns int (status)
**  		void *dns -- DNS service handle
**  		void *qh -- query handle to be canceled
*/

void
iprange_dns_set_query_cancel(IPRANGE *rbl, int (*func)(void *, void *))
{
	assert(rbl != NULL);

	iprange->iprange_dns_cancel = func;
}

/*
**  IPRANGE_DNS_SET_QUERY_WAITREPLY -- stores a pointer to wait for a DNS reply
**
**  Parameters:
**  	iprange -- IPRANGE library handle
**  	func -- function to use to wait for a reply
**
**  Return value:
**  	None.
**
**  Notes:
**  	"func" should match the following prototype:
**  		returns int (status)
**  		void *dns -- DNS service handle
**  		void *qh -- handle of query that has completed
**  		struct timeval *timeout -- how long to wait
**  		size_t *bytes -- bytes returned
**  		int *error -- error code returned
**  		int *dnssec -- DNSSEC status returned
*/

void
iprange_dns_set_query_waitreply(IPRANGE *rbl, int (*func)(void *, void *,
                                                          struct timeval *,
                                                          size_t *, int *,
                                                          int *))
{
	assert(rbl != NULL);

	iprange->iprange_dns_waitreply = func;
}

/*
**  IPRANGE_QUERY_CANCEL -- cancel an open query to the IPRANGE
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	qh -- query handle
**
**  Return value:
** 	IPRANGE_STAT_* -- as defined
*/

IPRANGE_STAT
iprange_query_cancel(IPRANGE *rbl, void *qh)
{
	struct iprange_query *rq;

	assert(rbl != NULL);
	assert(qh != NULL);

	rq = qh;

	iprange->iprange_dns_cancel(iprange->iprange_dns_service, rq->rq_qh);

	if (iprange->iprange_free != NULL)
		iprange->iprange_free(iprange->iprange_closure, rq);
	else
		free(rq);

	return IPRANGE_STAT_OK;
}

/*
**  IPRANGE_QUERY_START -- initiate a query to the IPRANGE for entries
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	query -- query string
**  	qh -- query handle (returned)
**
**  Return value:
** 	IPRANGE_STAT_* -- as defined
*/

IPRANGE_STAT
iprange_query_start(IPRANGE *rbl, u_char *query, void **qh)
{
	int status;
	struct iprange_query *rq;
	u_char rblquery[IPRANGE_MAXHOSTNAMELEN + 1];

	assert(rbl != NULL);
	assert(query != NULL);
	assert(qh != NULL);

	if (iprange->iprange_qroot[0] == '\0')
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "query root not set");
		return IPRANGE_STAT_INVALID;
	}

	snprintf(rblquery, sizeof rblquery, "%s.%s", query,
	         iprange->iprange_qroot);

	if (iprange->iprange_malloc != NULL)
	{
		rq = iprange->iprange_malloc(iprange->iprange_closure,
		                             sizeof(*rq));
	}
	else
	{
		rq = malloc(sizeof(*rq));
	}

	if (rq == NULL)
		return IPRANGE_STAT_NORESOURCE;

	memset(rq, '\0', sizeof *rq);

	status = iprange->iprange_dns_start(iprange->iprange_dns_service,
	                                    T_A, rblquery, rq->rq_buf,
	                                    sizeof rq->rq_buf, &rq->rq_qh);

	if (status == 0)
	{
		*qh = rq;
		return IPRANGE_STAT_OK;
	}
	else
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "unable to start query for '%s'", rblquery);
		return IPRANGE_STAT_DNSERROR;
	}
}

/*
**  IPRANGE_QUERY_CHECK -- check for a reply from an active query
**
**  Parameters:
**  	iprange -- IPRANGE handle, created by iprange_init()
**  	qh -- query handle (returned)
**  	timeout -- timeout
**  	res -- 32-bit buffer into which to write the result (can be NULL)
**
**  Return value:
** 	IPRANGE_STAT_* -- as defined
*/

IPRANGE_STAT
iprange_query_check(IPRANGE *rbl, void *qh, struct timeval *timeout,
                    uint32_t *res)
{
	int dnserr;
	int status;
	int n;
	int type;
	int class;
	int qdcount;
	int ancount;
	struct iprange_query *rq;
	u_char *cp;
	u_char *eom;
	u_char *found = NULL;
	HEADER hdr;
	u_char qname[IPRANGE_MAXHOSTNAMELEN + 1];

	assert(rbl != NULL);
	assert(qh != NULL);

	rq = qh;

	status = iprange->iprange_dns_waitreply(iprange->iprange_dns_service,
	                                rq->rq_qh, timeout, &rq->rq_anslen,
	                                &dnserr, NULL);

	if (status == IPRANGE_DNS_ERROR)
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "error during query");
		return IPRANGE_STAT_ERROR;
	}
	else if (status == IPRANGE_DNS_NOREPLY)
	{
		return IPRANGE_STAT_NOREPLY;
	}
	else if (status == IPRANGE_DNS_EXPIRED)
	{
		return IPRANGE_STAT_EXPIRED;
	}

	/* set up pointers */
	memcpy(&hdr, rq->rq_buf, sizeof hdr);
	cp = (u_char *) rq->rq_buf + HFIXEDSZ;
	eom = (u_char *) rq->rq_buf + rq->rq_anslen;

	/* skip over the name at the front of the answer */
	for (qdcount = ntohs((unsigned short) hdr.qdcount);
	     qdcount > 0;
	     qdcount--)
	{
		/* copy it first */
		(void) dn_expand((unsigned char *) rq->rq_buf, eom, cp,
		                 (char *) qname, sizeof qname);
 
		if ((n = dn_skipname(cp, eom)) < 0)
		{
			snprintf(iprange->iprange_error,
			         sizeof iprange->iprange_error,
			         "'%s' reply corrupt", qname);
			return IPRANGE_STAT_ERROR;
		}
		cp += n;

		/* extract the type and class */
		if (cp + INT16SZ + INT16SZ > eom)
		{
			snprintf(iprange->iprange_error,
			         sizeof iprange->iprange_error,
			         "'%s' reply corrupt", qname);
			return IPRANGE_STAT_ERROR;
		}
		GETSHORT(type, cp);
		GETSHORT(class, cp);
	}

	if (type != T_A || class != C_IN)
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "'%s' unexpected reply type/class", qname);
		return IPRANGE_STAT_ERROR;
	}

	/* if NXDOMAIN, return DKIM_STAT_NOKEY */
	if (hdr.rcode == NXDOMAIN)
		return IPRANGE_STAT_NOTFOUND;

	/* get the answer count */
	ancount = ntohs((unsigned short) hdr.ancount);
	if (ancount == 0)
		return IPRANGE_STAT_NOTFOUND;

	/*
	**  Extract the data from the first TXT answer.
	*/

	while (--ancount >= 0 && cp < eom)
	{
		/* grab the label, even though we know what we asked... */
		if ((n = dn_expand((unsigned char *) rq->rq_buf, eom, cp,
		                   (RES_UNC_T) qname, sizeof qname)) < 0)
		{
			snprintf(iprange->iprange_error,
			         sizeof iprange->iprange_error,
			         "'%s' reply corrupt", qname);
			return IPRANGE_STAT_ERROR;
		}
		/* ...and move past it */
		cp += n;

		/* extract the type and class */
		if (cp + INT16SZ + INT16SZ > eom)
		{
			snprintf(iprange->iprange_error,
			         sizeof iprange->iprange_error,
			         "'%s' reply corrupt", qname);
			return IPRANGE_STAT_ERROR;
		}

		GETSHORT(type, cp);
		GETSHORT(class, cp);

		/* skip the TTL */
		cp += INT32SZ;

		/* skip CNAME if found; assume it was resolved */
		if (type == T_CNAME)
		{
			char chost[IPRANGE_MAXHOSTNAMELEN + 1];

			n = dn_expand((u_char *) rq->rq_buf, eom, cp,
			              chost, IPRANGE_MAXHOSTNAMELEN);
			cp += n;
			continue;
		}
		else if (type == T_RRSIG)
		{
			/* get payload length */
			if (cp + INT16SZ > eom)
			{
				snprintf(iprange->iprange_error,
				         sizeof iprange->iprange_error,
				         "'%s' reply corrupt", qname);
				return IPRANGE_STAT_ERROR;
			}
			GETSHORT(n, cp);

			cp += n;

			continue;
		}
		else if (type != T_A)
		{
			snprintf(iprange->iprange_error,
			         sizeof iprange->iprange_error,
			         "'%s' unexpected reply type/class", qname);
			return IPRANGE_STAT_ERROR;
		}

		if (found != NULL)
		{
			snprintf(iprange->iprange_error,
			         sizeof iprange->iprange_error,
			         "multiple replies for '%s'", qname);
			return IPRANGE_STAT_ERROR;
		}

		/* remember where this one started */
		found = cp;

		/* get payload length */
		if (cp + INT16SZ > eom)
		{
			snprintf(iprange->iprange_error,
			         sizeof iprange->iprange_error,
			         "'%s' reply corrupt", qname);
			return IPRANGE_STAT_ERROR;
		}
		GETSHORT(n, cp);

		/* move forward for now */
		cp += n;
	}

	/* if ancount went below 0, there were no good records */
	if (found == NULL)
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "'%s' reply was unresolved CNAME", qname);
		return IPRANGE_STAT_ERROR;
	}

	/* come back to the one we found */
	cp = found;

	/* get payload length */
	if (cp + INT16SZ > eom)
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "'%s' reply corrupt", qname);
		return IPRANGE_STAT_ERROR;
	}

	GETSHORT(n, cp);
	if (n != sizeof(uint32_t))
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "'%s' reply corrupt", qname);
		return IPRANGE_STAT_ERROR;
	}

	if (cp + n > eom)
	{
		snprintf(iprange->iprange_error, sizeof iprange->iprange_error,
		         "'%s' reply corrupt", qname);
		return IPRANGE_STAT_ERROR;
	}

	/* extract the payload */
	if (res != NULL)
	{
		uint32_t addr;

		GETLONG(addr, cp);

		*res = addr;
	}

	return IPRANGE_STAT_FOUND;
}
